### PR TITLE
added hot reload to receptor config

### PR DESCRIPTION
**Pull request by resoluteCoder**

### CODE REVIEW

File: cmd/config.go [hunk 1]
```diff
@@ -157,6 +174,33 @@ func RunPhases(phase string, v reflect.Value) {
 	}
 }
 
+// ReloadServices iterates through key/values calling reload on applicable services.
+func ReloadServices(v reflect.Value) {
+	for i := 0; i < v.NumField(); i++ {
+		if reflect.Value.IsZero(v.Field(i)) {
+			continue
```
Comment (AaronH88) [inline cmd/config.go:181]: Add a comment here please, 
Suggestion:
"If the service is not part of the backends struct, do nothing"

File: cmd/config.go [hunk 2]
```diff
@@ -157,6 +174,33 @@ func RunPhases(phase string, v reflect.Value) {
 	}
 }
 
+// ReloadServices iterates through key/values calling reload on applicable services.
+func ReloadServices(v reflect.Value) {
+	for i := 0; i < v.NumField(); i++ {
+		if reflect.Value.IsZero(v.Field(i)) {
```
Comment (AaronH88) [inline cmd/config.go:181]: Add a comment here please,
Suggestion:
"If the service is not initialised, skip"

File: cmd/config.go [hunk 3]
```diff
@@ -157,6 +174,33 @@ func RunPhases(phase string, v reflect.Value) {
...
+			continue
+		}
+
+		var err error
+		switch v.Field(i).Kind() {
+		case reflect.Slice:
+			for j := 0; j < v.Field(i).Len(); j++ {
```
Comment (AaronH88) [inline cmd/config.go:189]: Add a comment here please,
Suggestion:
"Iterate over all the type fields"

File: cmd/config.go [hunk 4]
```diff
@@ -157,6 +174,33 @@ func RunPhases(phase string, v reflect.Value) {
...
+				serviceItem := v.Field(i).Index(j).Interface()
+				switch c := serviceItem.(type) {
+				case Reloader:
```
Comment (AaronH88) [inline cmd/config.go:194]: Add a comment here please,
Suggestion:
"Check to see if the selected type field is of type Reloader interface, meaning reload can be called"

File: cmd/config.go [hunk 5]
```diff
@@ -157,6 +174,33 @@ func RunPhases(phase string, v reflect.Value) {
...
+					err = c.Reload()
+					if err != nil {
+						PrintPhaseErrorMessage(v.Type().Name(), "reload", err)
+					}
+				default:
```
Comment (AaronH88) [inline cmd/config.go:200]: Add a comment here please,
Suggestion:
"If the service does not implement the Reloader interface, do nothing"

**Review by AaronH88 — COMMENTED** [review]
Overall looking good, 
I would like a few more comments around the reflection function to make it quicker to read once we revisit later

The functionality looks good to me,
@thom-at-redhat and @matoval can you look also please

File: cmd/root.go [hunk 6]
```diff
@@ -6,15 +6,20 @@ import (
...
 
 var (
-	cfgFile string
-	version bool
+	cfgFile       string
+	version       bool
+	backendConfig *BackendConfig
 )
 
+var FindMe = true
```
Comment (matoval) [inline cmd/root.go]: What is this var used for?

File: cmd/root.go [hunk 7]
```diff
@@ -78,8 +60,85 @@ func initConfig() {
...
+		// allows to skip empty first config
+		isEmpty := isConfigEmpty(reflect.ValueOf(*newConfig))
+		if isEmpty {
+			return
+		}
+
+		SetBackendConfigDefaults(newConfig)
+
+		isEqual := reflect.DeepEqual(*backendConfig, *newConfig)
+		if !isEqual {
```
Comment (matoval) [inline cmd/root.go:79]: Should we add an else that alerts the user that the config change wasn't reloadable with a log?

File: cmd/root.go [hunk 8]
```diff
@@ -78,8 +60,85 @@ func initConfig() {
 
 	viper.AutomaticEnv()
 
+	viper.OnConfigChange(func(e fsnotify.Event) {
+		fmt.Printf("Config file changed: %s\n", e.Name)
```
Comment (matoval) [inline cmd/root.go]: Could this be a log instead of a print?

**Review by matoval — COMMENTED** [review]
Looks great, I have a few questions

**Review by matoval — APPROVED** [review]

**Pull request merged**